callback functions - settingsChangedCallback and packetReceivedCallback

kayno opened the pull request:
Hi @SwiCago 

Thanks for merging #2! With that out of the way, I would like to slowly work through some of the other ideas for improvements (including some code doco) that I have.

This pull request is for some callback functions, to allow users of the library to code some action in their sketches whenever certain events happen in the library. The two events for callbacks I have in mind are:

- whenever the heatpump's settings are changed (for example if someone presses a button on the remote control, the sketch will be able to easily execute some code
- whenever a valid (checksum passes) packet is received from the heatpump (this one is mostly for debugging - it creates an wasy way for the user's sketch to get every packet that is received from the heatpump, without putting any debug code in the library itself.

The single commit in this pull request has:

- added two callback functions: one for when the heatpump settings change and the other (for debugging purposes) when a packet is received by the heatpump, which can be used as follows:

`void setup() {`
  `hp.setSettingsChangedCallback(hpSettingsChanged);`
  `hp.setPacketReceivedCallback(hpPacketReceived);`
`}`

`void hpSettingsChanged() {`
  `// settings have changed - do something...`
`}`

`void hpPacketReceived(byte* packet, unsigned int length) {`
  `// packet received from heatpump - do something...`
`}`

- updated mitsubishi_heatpump_mqtt_esp8266 example to use both callbacks. I will start to use this to debug other data in the packet, for example the isave mode where temperature can be set as low as 10

Please review, and I am happy to discuss more if needed :)

Thanks :)
kayno

SwiCago commented:
Hi Kayno, this is a good idea. In my original code I was also keeping the packets, so I could display them into a webpage for review. I was debugging an issue and didn't see my error until I saw the raw HEX.
I am merging this addition, I like it.

SwiCago commented:
Kayno, have a question for you. I the master I had a few day ago the cpp had the following
void HeatPump::setSettings(String settings[]) {
  setPowerSetting(settings[0]);
  setModeSetting(settings[1]);
  setTemperature(settings[2]);
  setFanSpeed(settings[3]);
  setVaneSetting(settings[4]);
  setDirectionSetting(settings[5]);
}

Any reason why this as changed to 
void HeatPump::setSettings(heatpumpSettings settings) {
  HeatPump::setPowerSetting(settings.power);
  HeatPump::setModeSetting(settings.mode);
  HeatPump::setTemperature(settings.temperature);
  HeatPump::setFanSpeed(settings.fan);
  HeatPump::setVaneSetting(settings.vane);
  HeatPump::setWideVaneSetting(settings.wideVane);
}
and not to 
void HeatPump::setSettings(heatpumpSettings settings) {
  setPowerSetting(settings.power);
  setModeSetting(settings.mode);
  setTemperature(settings.temperature);
  setFanSpeed(settings.fan);
  setVaneSetting(settings.vane);
  setWideVaneSetting(settings.wideVane);
}

The class does not need to reference itself as HeatPump:: when calling methods from within itself. Let me know if we can change it to the above?

kayno commented:
@SwiCago - I think originally the addition of "HeatPump::" came from unreality, which i forked and changed and moved back to yours (#2). See the original commit from unreality here: https://github.com/unreality/HeatPump/commit/555662a8484f19dd10a47cee14aeb03081f034d9#diff-aa5d1dbc2ff977e82fbccafda04a670cR87

I am a bit rusty on c++ and just carried it through. Happy for either of us to change it back - I can update and test it on my heatpump tonight if you like.

It would be good to develop a test sketch that we can use to test changes to the library, e.g. one that calls each set function, calls each get, validates that updates work, etc. I will add that to my TODO list!

BTW - thanks for the merge!

SwiCago commented:
@kayno I am currently away for the weekend, but I know it works w/o the HeatPump:: prefix. Give it a try and put it in your next round of updates. If you have a secpnd, add the alias method loop that will just call sync 
Thanks for all great aditions you have made. The callbacks was a great idea. 
One question for you, what are you using for automation? Openhab or something custom? In my case I have ^
6 zone hydronic and heat pump, so will most likely do something custom to be fully automatic. But am interested in what others are doing
